src/client.js

```diff
@@ -0,0 +1,229 @@
+    // maps a cabal-core instance to a CabalDetails object
+    this.cabals = new Map()
+    this.currentCabal = null
+    this.config = opts.config
+    this.maxFeeds = opts.maxFeeds || 1000
+
+    let cabalDnsOpts = {
+        hashRegex: /^[0-9a-f]{64}?$/i,
+        recordName: 'cabal',
+        protocolRegex: /^cabal:\/\/([0-9a-f]{64})/i,
+        txtRegex: /^"?cabalkey=([0-9a-f]{64})"?$/i,
+    }
+    // also takes opts.persistentCache which has a read and write function
+    //   read: async function ()   // aka cache lookup function
+    //   write: async function ()  // aka cache write function
+    if (opts.persistentCache) cabalDnsOpts.persistentCache = opts.persistentCache
+    this.cabalDns = DatDns(cabalDnsOpts)
+  }
+
+  static getDatabaseVersion () {
+    return Cabal.databaseVersion
+  }
+
+  static scrub (key) {
+    return key.replace('cabal://', '').replace('cbl://', '').replace('dat://', '').replace(/\//g, '')
+  }
+
+  resolveName (name, cb) {
+      return this.cabalDns.resolveName(name).then((key) => { 
+          if (!cb) return Client.scrub(key)
+          cb(Client.scrub(key)) 
+      })
+  }
+
+  createCabal () {
+    return this.addCabal(crypto.keyPair().publicKey.toString('hex'))
+  }
+
+  addCabal (key) {
+      let cabalPromise
+      if (typeof key === 'string') {
+          cabalPromise = this.resolveName(key).then((resolvedKey) => {
+            const {temp, dbdir} = this.config
+            const storage = temp ? ram : dbdir + resolvedKey
+            if (!temp) try { mkdirp.sync(path.join(dbdir, resolvedKey, 'views')) } catch (e) {}
+            var db = temp ? memdb() : level(path.join(dbdir, resolvedKey, 'views'))
+            var cabal = Cabal(storage, resolvedKey, {db: db, maxFeeds: this.maxFeeds})
+            this._keyToCabal[resolvedKey] = cabal
+            return cabal
+          })
+      } else {
+          cabalPromise = new Promise((res, rej) => {
+            // a cabal instance was passed in
+            var cabal = key
+            this._keyToCabal[cabal.key] = cabal
+            res(cabal)
+          })
+      }
+      return new Promise((res, rej) => {
+          cabalPromise.then((cabal) => {
+              if (!this.currentCabal) {
+                  this.currentCabal = cabal
+              }
+              cabal.ready(() => {
+                  const details = new CabalDetails(cabal)
+                  this.cabals.set(cabal, details)
+                  cabal.swarm()
+                  this.getCurrentCabal()._emitUpdate()
+                  res(details)
+              })
+          })
+    })
+  }
+
+  focusCabal (key) {
+    const cabal = this._coerceToCabal(key)
+    if (!cabal) {
+      return false
+    }
+    this.currentCabal = cabal
+    return this.cabalToDetails(cabal)
+  }
+
+  removeCabal (key) {
+    const cabal = this._coerceToCabal(key)
+    if (!cabal) {
+      return false
+    }
+
+    const details = this.cabalToDetails(cabal)
+    details._destroy()
```

Comment:
Are the cabal-core and the swarm also being terminated? Otherwise it'll continue peering and indexing.

(If this is non-trivial, we can make an issue & a future PR!)

Reply:
they're not, let's create an issue!

Reply:
issue created: https://github.com/cabal-club/cabal-client/issues/12